File: samples/js/text_generation/structural_tags_generation.js
// Copyright (C) 2025 Intel Corporation
// SPDX-License-Identifier: Apache-2.0

import readline from 'readline';
import { z } from 'zod';
import { LLMPipeline, StreamingStatus, StructuredOutputConfig } from 'openvino-genai-node';
import { serialize_json } from './helper.js';

const getWeatherTool = {
    name: "get_weather",
    schema: z.object({
        city: z.string().describe("City name"),
        country: z.string().describe("Country name"),
        date: z.string().regex(/2\d{3}-[0-1]\d-[0-3]\d/).describe("Date in YYYY-MM-DD format")
    }),
};

const getCurrencyExchangeTool = {
    name: "get_currency_exchange",
    schema: z.object({
        from_currency: z.string().describe("Currency to convert from"),
        to_currency: z.string().describe("Currency to convert to"),
        amount: z.number().describe("Amount to convert")
    }),
};

const tools = [getWeatherTool, getCurrencyExchangeTool];

const sysMessage = "You are a helpful assistant that can provide weather information and currency exchange rates. "
    + `Today is ${new Date().toISOString().split('T')[0]}. `
    + "You can respond in natural language, always start your answer with appropriate greeting, "
    + "If you need additional information to respond you can request it by calling particular tool with structured JSON. "
    + `You can use the following tools:
${tools.map(tool => `<function_name=\"${tool.name}\">, arguments=${serialize_json(tool.schema.keyof().options)}`).join('\n')}
Please, only use the following format for tool calling in your responses:
<function=\"function_name\">{"argument1": "value1", ...}</function>
Use the tool name and arguments as defined in the tool schema.
If you don't know the answer, just say that you don't know, but try to call the tool if it helps to answer the question.`;

const functionPattern = /<function="([^"]+)">(.*?)<\/function>/gs;

/** Parse the tool response from the model output.
    The response should be in the format:
    <function="function_name">{"argument1": "value1", ...}</function>
    */
function parseToolsFromResponse(response) {
    const matches = response.matchAll(functionPattern);
    return Array.from(matches).map(match => {
        const toolName = match[1];
        const args = JSON.parse(match[2]);
        return { toolName, args };
    });
}

function streamer(subword) {
    process.stdout.write(subword);
    return StreamingStatus.RUNNING;
}

async function main() {
    const defaultPrompt = "What is the weather in London today and in Paris yesterday, and how many pounds can I get for 100 euros?";

    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    const modelDir = process.argv[2];
    if (!modelDir) {
        console.error('Please provide the path to the model directory as the first argument.');
        process.exit(1);
    }

    const prompt = process.argv[3] || defaultPrompt;

    rl.close();

    const device = "CPU"; // GPU can be used as well
    const pipe = await LLMPipeline(modelDir, device);

    console.log(`User prompt: ${prompt} `);

    for (const useStructuralTags of [false, true]) {
        console.log("=".repeat(80));
        console.log(`${useStructuralTags ? "Using structural tags" : "Using no structural tags"} `.padStart(40).padEnd(80));
        console.log("=".repeat(80));

        const generation_config = {};
        generation_config.return_decoded_results = true;
        generation_config.max_new_tokens = 300;

        await pipe.startChat(sysMessage);
        if (useStructuralTags) {
            generation_config.structured_output_config = {
                structural_tags_config: StructuredOutputConfig.TriggeredTags({
                    tags: tools.map(tool => StructuredOutputConfig.Tag({
                        begin: `<function=\"${tool.name}\">`,
                        content: StructuredOutputConfig.JSONSchema(serialize_json(z.toJSONSchema(tool.schema))),
                        end: "</function>"
                    })),
                    triggers: ["<function="]
                })
            };
        };
        generation_config.do_sample = true;

        const response = await pipe.generate(prompt, generation_config, streamer);
        await pipe.finishChat();
        console.log("\n" + "-".repeat(80));

        console.log("Correct tool calls by the model:");
        console.log(parseToolsFromResponse(response.toString()));
    }
}

main();

File: samples/python/text_generation/structural_tags_generation.py
#!/usr/bin/env python3
# Copyright (C) 2025 Intel Corporation
# SPDX-License-Identifier: Apache-2.0

import argparse
import json
import re
from datetime import datetime
from pprint import pprint
from typing import ClassVar

from openvino_genai import (
    GenerationConfig,
    LLMPipeline,
    StreamingStatus,
)
from openvino_genai import (
    StructuredOutputConfig as SOC,
)
from pydantic import BaseModel, Field


class ToolRequest(BaseModel):
    @classmethod
    def string_representation(cls) -> str:
        return f'<function_name="{cls.get_name()}">, arguments={list(cls.model_fields)}'

    @classmethod
    def get_name(cls) -> str:
        return cls._name


class WeatherRequest(ToolRequest):
    _name: ClassVar[str] = "get_weather"

    city: str = Field(description="City name")
    country: str = Field(description="Country name")
    date: str = Field(pattern=r"2\d\d\d-[0-1]\d-[0-3]\d", description="Date in YYYY-MM-DD format")


class CurrencyExchangeRequest(ToolRequest):
    _name: ClassVar[str] = "get_currency_exchange"

    from_currency: str = Field(description="Currency to convert from")
    to_currency: str = Field(description="Currency to convert to")
    amount: float = Field(description="Amount to convert")


tools = {tool.get_name(): tool for tool in [WeatherRequest, CurrencyExchangeRequest]}

new_line = "\n"  # to use inside f-string
sys_message = (
    "You are a helpful assistant that can provide weather information and currency exchange rates. "
    f"Today is {datetime.today().strftime('%Y-%m-%d')}. "  # Use the current date in the system message in YYYY-MM-DD format
    "You can respond in natural language, always start your answer with appropriate greeting, "
    "If you need additional information to respond you can request it by calling particular tool with structured JSON. "
    "You can use the following tools:\n"
    f"{new_line.join([tool.string_representation() for tool in tools.values()])}\n"
    "Please, only use the following format for tool calling in your responses:\n"
    '<function="function_name">'
    '{"argument1": "value1", ...}'
    "</function>\n"
    "Use the tool name and arguments as defined in the tool schema.\n"
    "If you don't know the answer, just say that you don't know, but try to call the tool if it helps to answer the question.\n"
)

function_pattern = r'<function="([^"]+)">({.*?})</function>'
function_pattern = re.compile(function_pattern, re.DOTALL)


def parse_tools_from_response(response: str) -> list[ToolRequest]:
    """
    Parse the tool response from the model output.
    The response should be in the format:
    <function="function_name">{"argument1": "value1", ...}</function>
    """
    matches = re.finditer(function_pattern, response)
    return [tools.get(match.group(1)).model_validate_json(match.group(2)) for match in matches]


def streamer(subword):
    print(subword, end="", flush=True)
    return StreamingStatus.RUNNING


def main():
    default_prompt = (
        "What is the weather in London today and in Paris yesterday, and how many pounds can I get for 100 euros?"
    )

    description = (
        "This script demonstrates how to use OpenVINO GenAI with structured tags to generate responses "
        "that include tool calls. It uses a simple LLM pipeline to generate a response based on the provided prompt, "
        "and it parses the tool calls from the response. Available tools are weather and currency exchange."
    )
    parser = argparse.ArgumentParser(description=description)
    parser.add_argument(
        "model_dir",
        help="Path to the model directory. It should contain the OpenVINO model files.",
    )
    parser.add_argument(
        "--prompt",
        type=str,
        default=default_prompt,
        help="Prompt to generate the response.",
    )
    args = parser.parse_args()

    device = "CPU"  # GPU can be used as well
    pipe = LLMPipeline(args.model_dir, device)

    print(f"User prompt: {args.prompt}")

    for use_structural_tags in [False, True]:
        print("=" * 80)
        print(f"{'Using structural tags' if use_structural_tags else 'Using no structural tags':^80}")
        print("=" * 80)
        config = GenerationConfig()
        config.max_new_tokens = 300

        pipe.start_chat(sys_message)
        if use_structural_tags:
            config.structured_output_config = SOC(
                structural_tags_config=SOC.TriggeredTags(
                    triggers=["<function="],
                    tags=[
                        SOC.Tag(
                            begin=f'<function="{name}">',
                            content=SOC.JSONSchema(json.dumps(tool.model_json_schema())),
                            end="</function>",
                        )
                        for name, tool in tools.items()
                    ],
                )
            )
            config.do_sample = True
        response = pipe.generate(args.prompt, config, streamer=streamer)
        pipe.finish_chat()
        print("\n" + "-" * 80)

        print("Correct tool calls by the model:")
        pprint(parse_tools_from_response(response))


if "__main__" == __name__:
    main()
